### PR TITLE
Optimizations

### PR DESCRIPTION
**Pull request by molaxx**

### CODE REVIEW

File: computation_graph/run.py [hunk 1]
```diff
@@ -281,93 +263,137 @@ def process_node(
...
-            )
-        ),
-        immutables.Map(),
-    )
-
-
-def _combine_inputs_with_edges(
-    inputs: _NodeToResults,
-) -> Callable[[base_types.GraphType], base_types.GraphType]:
+def _future_edge_to_regular_with_placeholder(
```
Comment (itayzit) [inline computation_graph/run.py:266]:
```suggestion
def _future_edge_to_regular_edges_with_placeholder(
```

File: computation_graph/run.py [hunk 2]
```diff
@@ -281,93 +263,137 @@ def process_node(
...
 def _to_callable_with_side_effect_for_single_and_multiple(
     single_node_side_effect: _SingleNodeSideEffect,
     all_nodes_side_effect: Callable,
     edges: base_types.GraphType,
-    handled_exceptions: FrozenSet[Type[Exception]],
+    handled_exceptions: Tuple[Type[Exception], ...],
 ) -> Callable[[_NodeToResults, _NodeToResults], _NodeToResults]:
+    future_sources = opt_gamla.compose_left(
+        opt_gamla.filter(base_types.edge_is_future),
+        # Assumption, no "args" edges as sources
```
Comment (itayzit) [inline computation_graph/run.py:287]:
```suggestion
        # At this stage we assume there are no "args" edges as sources
```

File: computation_graph/run.py [hunk 3]
```diff
@@ -281,93 +263,137 @@ def process_node(
...
+    future_source_to_placeholder = opt_gamla.compose_left(
+        opt_gamla.map(
+            opt_gamla.pair_right(
+                opt_gamla.compose_left(
+                    gamla.attrgetter("name"), graph.make_source_with_name
+                )
+            )
+        ),
+        gamla.frozendict,
+    )(future_sources)
```
Comment (itayzit) [inline computation_graph/run.py:300]: use a pipe/factor out to a function

File: computation_graph/run.py [hunk 4]
```diff
@@ -281,93 +263,137 @@ def process_node(
...
 
 
 def _to_callable_with_side_effect_for_single_and_multiple(
     single_node_side_effect: _SingleNodeSideEffect,
     all_nodes_side_effect: Callable,
     edges: base_types.GraphType,
-    handled_exceptions: FrozenSet[Type[Exception]],
+    handled_exceptions: Tuple[Type[Exception], ...],
 ) -> Callable[[_NodeToResults, _NodeToResults], _NodeToResults]:
+    future_sources = opt_gamla.compose_left(
```
Comment (itayzit) [inline computation_graph/run.py:285]: use a pipe/factor out to a function

File: computation_graph/run.py [hunk 5]
```diff
@@ -281,93 +263,137 @@ def process_node(
...
             )
 
-    return (_async_graph_reducer if is_async else _graph_reducer)(
-        gamla.compose(
-            final_runner, lambda inputs: _combine_inputs_with_edges(inputs)(edges)
-        )
+    return (_async_graph_reducer if is_async else _graph_reducer)(final_runner)
+
+
+def _make_reduce_layers(
```
Comment (itayzit) [inline computation_graph/run.py:364]: should this be:
```suggestion
def _make_reduce_layer(
```

**Review by itayzit — APPROVED** [review]
🚀

File: computation_graph/run.py (hunk 5 above)
Comment (molaxx) [reply computation_graph/run.py:364]: this reduce layers, not a single one

**Pull request merged**